pkg/coalescer/coalescer.go

```diff
@@ -0,0 +1,164 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package coalescer
+
+import (
+	"time"
+
+	"k8s.io/klog/v2"
+)
+
+// A coalescer interface to combine multiple requests made over a period of time into a single request
+//
+// When a request is received that matches an existing in-flight request, the coalescer will attempt to
+// merge that request into the existing request pool using the provided mergeFunction
+//
+// When the delay on the request expires (determined by the time the first request comes in), the merged
+// input is passed to the execution function, and the result to all waiting callers (those that were
+// not rejected during the merge step)
+type Coalescer[InputType comparable, ResultType any] interface {
+	// Coalesces a given input
+	// key = only requests with this same key will be coalesced (such as volume ID)
+	// input = input to merge with other inputs
+	// It is NOT guaranteed all callers receive the same result (for example, if
+	// an input fails to merge, only that caller will receive an error)
+	Coalesce(key string, input InputType) (ResultType, error)
+}
+
+// Creates a new coalescer and immediately begins processing requests on a goroutine
+// delay = the time to wait for other requests to coalesce before executing
+// mergeFunction = a function to merge a new input with the existing inputs
+// (should return an error if the new input cannot be combined with the existing inputs,
+// otherwise return the new merged input)
+// executeFunction = the function to call when the delay expires
+func New[InputType comparable, ResultType any](delay time.Duration,
+	mergeFunction func(input InputType, existing InputType) (InputType, error),
+	executeFunction func(key string, input InputType) (ResultType, error),
+) Coalescer[InputType, ResultType] {
+	c := coalescer[InputType, ResultType]{
+		delay:           delay,
+		mergeFunction:   mergeFunction,
+		executeFunction: executeFunction,
+		inputChannel:    make(chan newInput[InputType, ResultType]),
+		timerChannel:    make(chan string),
+		pendingInputs:   make(map[string]pendingInput[InputType, ResultType]),
+	}
+
+	go c.coalescerThread()
+	return &c
+}
+
+// Type to store a result or error in channels
+type result[ResultType any] struct {
+	result ResultType
+	err    error
+}
+
+// Type to send inputs from Coalesce() to coalscerThread() via channel
```

Comment:
nit: typo `coalscerThread` ->  `coalescerThread`

Reply:
Fixed in next revision